### PR TITLE
Creating react-focus package and copying FocusZone code there

### PR DESCRIPTION
**Pull request by dzearing**
This PR introduces an initial breakout of the focus components into a `react-focus` package.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/11765)

### CODE REVIEW

File: packages/react-focus/package.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
+{
+  "name": "@fluentui/react-focus",
+  "version": "0.0.1",
```
Comment (ecraig12345) [inline packages/react-focus/package.json]: Is this intentional? create-package usually does 0.1.0 to avoid the weirdness of 0.0.x semver.

File: packages/react-focus/package.json [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    "@types/webpack-env": "1.13.9",
+    "@uifabric/build": "^7.0.0",
+    "@uifabric/example-app-base": "^7.11.1",
+    "@uifabric/jest-serializer-merge-styles": "^7.0.6",
+    "@uifabric/prettier-rules": "^7.0.4",
+    "@uifabric/tslint-rules": "^7.0.5",
+    "enzyme": "~3.9.0",
+    "enzyme-adapter-react-16": "^1.2.0",
+    "es6-weak-map": "^2.0.2",
+    "office-ui-fabric-react": "^7.81.0",
```
Comment (ecraig12345) [inline packages/react-focus/package.json]: Should this be depending on OUFR at all? Will that create a circular link in graphs?

File: packages/react-focus/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# @fluentui/react-focus
+
+**React Focus components for [Fluent UI](https://fluent.microsoft.com/)**
```
Comment (ecraig12345) [inline packages/react-focus/README.md]: Can we update the create-package template to do this (if it is what we want for other packages too)?

File: packages/react-focus/package.json [hunk 4]
```diff
@@ -0,0 +1,61 @@
+{
+  "name": "@fluentui/react-focus",
```
Comment (ecraig12345) [inline packages/react-focus/package.json]: Will beachball support publishing to two scopes @kenotron ?

File: packages/react-focus/src/components/FocusZone/FocusZone.tsx [hunk 5]
```diff
@@ -0,0 +1,1201 @@
+import * as React from 'react';
```
Comment (ecraig12345) [inline packages/react-focus/src/components/FocusZone/FocusZone.tsx]: Are these files just a copy of the ones in OUFR?

File: packages/react-focus/src/components/FocusZone/FocusZonePage.tsx [hunk 6]
```diff
@@ -0,0 +1,73 @@
+import * as React from 'react';
```
Comment (ecraig12345) [inline packages/react-focus/src/components/FocusZone/FocusZonePage.tsx]: We don't have to change this now but we should consider whether we actually want these to live inside the relevant package (which will definitely create a circular dep just for example stuff if OUFR eventually depends on this new package) or inside fabric-website or something. 

If we made that change everywhere, it would also let us reduce the size of published packages. It would be a big change but definitely something that could be worth doing (and which I'd be interested in doing).

File: packages/react-focus/package.json [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    "react": "16.8.6",
+    "react-app-polyfill": "~1.0.1",
+    "react-dom": "16.8.6",
+    "react-test-renderer": "^16.3.0"
+  },
+  "dependencies": {
+    "@uifabric/set-version": "^7.0.2",
+    "@uifabric/utilities": "^7.11.0",
+    "@uifabric/merge-styles": "^7.8.1",
+    "tslib": "^1.7.1"
```
Comment (JasonGore) [inline packages/react-focus/package.json]: This has to be `^1.10.0` to prevent `__spreadArray` exceptions per #11739. Did this come from `create-package` or a template? If so please let me know, it has to be fixed.

File: yarn.lock [hunk 8]
```diff
@@ -14404,7 +14404,7 @@ ts-loader@^4.1.0:
     micromatch "^3.1.4"
     semver "^5.0.1"
 
-tslib@^1.10.0, tslib@^1.8.0, tslib@^1.8.1, tslib@^1.9.0:
+tslib@^1.10.0, tslib@^1.7.1, tslib@^1.8.0, tslib@^1.8.1, tslib@^1.9.0:
```
Comment (JasonGore) [inline yarn.lock]: Per my other comment, we shouldn't see any changes in this file.

File: packages/react-focus/package.json (hunk 1 above)
Comment (dzearing) [reply packages/react-focus/package.json]: i wasn't sure what the weirdness was. I can certainly revert but what's the weirdness?

File: packages/react-focus/package.json (hunk 2 above)
Comment (dzearing) [reply packages/react-focus/package.json]: no it should not, but i'm a little unsure about the docs solution.

The docs refer to components within oufr. So i will need to figure out a way to address keeping the examples somewhere else.

File: packages/react-focus/package.json (hunk 4 above)
Comment (dzearing) [reply packages/react-focus/package.json]: i don't see this being a problem.

File: packages/react-focus/src/components/FocusZone/FocusZone.tsx (hunk 5 above)
Comment (dzearing) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: yes, for now.

Comment (dzearing) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: This is just a scaffolding package. Once it is in a good shape, we can do a separate PR to move the other one to new location.

File: packages/react-focus/src/components/FocusZone/FocusZonePage.tsx (hunk 6 above)
Comment (dzearing) [reply packages/react-focus/src/components/FocusZone/FocusZonePage.tsx]: yes! I was thinking... should these be in umm, react-focus-docs

File: packages/react-focus/package.json (hunk 7 above)
Comment (dzearing) [reply packages/react-focus/package.json]: Yes I believe this was from the template.

File: yarn.lock (hunk 8 above)
Comment (dzearing) [reply yarn.lock]: weird!

File: packages/react-focus/package.json (hunk 1 above)
Comment (ecraig12345) [reply packages/react-focus/package.json]: I think it's mainly that `^0.0.1` will only pick up **specifically** that version (essentially `^` or `~` is ignored).

Also IIRC `^` behaves like `~` for `0.(>0).x`, but that's not quite as surprising.

File: packages/react-focus/package.json (hunk 2 above)
Comment (ecraig12345) [reply packages/react-focus/package.json]: Yeah we should potentially consider moving the examples to fabric-website or maybe even a dedicated examples package. That would fix a substantial number of circular dep issues.

File: packages/react-focus/src/components/FocusZone/FocusZonePage.tsx (hunk 6 above)
Comment (ecraig12345) [reply packages/react-focus/src/components/FocusZone/FocusZonePage.tsx]: A dedicated docs package for each published package seems like a bit much. Though if you put them all in the same package, once we multiple components by the same name, you have to figure out how to distinguish them, so...not sure of the best approach here.

File: yarn.lock (hunk 8 above)
Comment (JasonGore) [reply yarn.lock]: Well, this change is caused by the tslib `^1.7.1` usage above. When you change it to `^1.10.0`, this file should revert to having no changes. :)

File: packages/react-focus/package.json (hunk 7 above)
Comment (JasonGore) [reply packages/react-focus/package.json]: I ran `yarn create-package` and it did use tslib `^1.10.0` correctly, maybe this was created before my PR was merged. If you used `create-package` and ran into any issues, let me know and I'll queue them up. I'll have to update it anyways per upcoming CI changes (project references.)

File: packages/react-focus/package.json (hunk 2 above)
Comment (khmakoto) [reply packages/react-focus/package.json]: +1 for separating examples

File: scripts/just.config.ts [hunk 9]
```diff
@@ -1,29 +1,31 @@
...
-import prettier from './tasks/prettier';
-import bundleSizeCollect from './tasks/bundle-size-collect';
-import checkForModifiedFiles from './tasks/check-for-modified-files';
-import generateVersionFiles from './tasks/generate-version-files';
-import generatePackageManifestTask from './tasks/generate-package-manifest';
-import { postprocessTask } from './tasks/postprocess';
-import { postprocessAmdTask } from './tasks/postprocess-amd';
-import { postprocessCommonjsTask } from './tasks/postprocess-commonjs';
-import { startStorybookTask, buildStorybookTask } from './tasks/storybookTask';
+// @ts-check
```
Comment (kenotron) [inline scripts/just.config.ts]: Is there a reason this is turned back from typescript syntax back to javascript?

Comment (ecraig12345) [reply scripts/just.config.ts]: The change to TS syntax was actually just in a particular commit of this PR, not in master. It broke some other stuff, so I turned it back for now for simplicity.

**Comment by ecraig12345** [comment]
Made some updates to this PR after discussion with @kenotron:
- Delete OUFR FocusZone and re-export the one from react-focus
- The examples and docs stay in OUFR for now to avoid a circular dependency
- Addressed outstanding comments aside from ones pertaining to create-package and a separate long-term example location, which will be handled separately
- Add API Extractor to react-focus
- Update api-docs and tsx-editor to support react-focus

File: yarn.lock (hunk 8 above)
Comment (ecraig12345) [reply yarn.lock]: I fixed this.

File: packages/react-focus/src/components/FocusZone/FocusZonePage.tsx (hunk 6 above)
Comment (ecraig12345) [reply packages/react-focus/src/components/FocusZone/FocusZonePage.tsx]: Thinking about this again, the simplest solution would be to have the examples in a docs or website package, with sub-folders per package (which solves the issue of how to have examples for both e.g. OUFR button and experiments button). Definitely not happening in this PR though.

File: packages/react-focus/package.json (hunk 4 above)
Comment (ecraig12345) [reply packages/react-focus/package.json]: Before this merges, @kenotron needs to make sure that the account we use for publishing has permission to publish to `@fluentui` as well.

File: packages/react-focus/package.json (hunk 2 above)
Comment (ecraig12345) [reply packages/react-focus/package.json]: Left the examples in OUFR for now to avoid the circular dep and will move elsewhere long-term

File: packages/react-focus/src/components/FocusZone/FocusZone.test.tsx [hunk 10]
```diff
@@ -2,11 +2,11 @@ import * as React from 'react';
 import * as ReactDOM from 'react-dom';
 import * as renderer from 'react-test-renderer';
 import * as ReactTestUtils from 'react-dom/test-utils';
-import { setRTL, KeyCodes } from '../../Utilities';
+import { setRTL, KeyCodes } from '@uifabric/utilities';
```
Comment (JasonGore) [inline packages/react-focus/src/components/FocusZone/FocusZone.test.tsx]: 👍

File: packages/react-focus/package.json (hunk 4 above)
Comment (kenotron) [reply packages/react-focus/package.json]: I need the owner of uifabricteam on npm js.org to accept my invitation. Likely @dzearing?

**Comment by ecraig12345** [comment]
The mysterious test failure was due to a missing path mapping causing ComponentExamples.test.tsx to fail, but it was really hard to debug because the test was eating error messages (displaying a custom error instead of the real error). That has now been substantially improved, though I'm still not sure why the error messages weren't showing up in the CI build when they did show up locally.

**Comment by ecraig12345** [comment]
ComponentExamples stuff has been split out into #11849

**Comment by ecraig12345** [comment]
Tests are failing with this error, **on both this and an unrelated PR**...not sure what's going on.

```
office-ui-fabric-react:   ● DatePicker › should reflect the correct date in the input field when selecting a value
office-ui-fabric-react:     TypeError: Cannot read property '__reactInternalInstance$lwv38jf7tt' of null
office-ui-fabric-react: 
office-ui-fabric-react:       226 |     datePicker.setState({ isDatePickerShown: true });
office-ui-fabric-react:       227 |     const todayButton = document.querySelector('.ms-DatePicker-day--today') as HTMLButtonElement;
office-ui-fabric-react:     > 228 |     ReactTestUtils.Simulate.click(todayButton);
office-ui-fabric-react:           |                             ^
office-ui-fabric-react:       229 | 
office-ui-fabric-react:       230 |     const selectedDate = datePicker
office-ui-fabric-react:       231 |       .find('input')
office-ui-fabric-react: 
office-ui-fabric-react:       at getInstanceFromNode$1 (../../node_modules/react-dom/cjs/react-dom.development.js:888:18)
office-ui-fabric-react:       at Object.click (../../node_modules/react-dom/cjs/react-dom-test-utils.development.js:1209:22)
office-ui-fabric-react:       at Object.<anonymous> (src/components/DatePicker/DatePicker.test.tsx:228:29)
office-ui-fabric-react:   ● DatePicker › should reflect the correct date in the input field when selecting a value and a different format is given
office-ui-fabric-react:     TypeError: Cannot read property '__reactInternalInstance$lwv38jf7tt' of null
office-ui-fabric-react: 
office-ui-fabric-react:       250 |     datePicker.setState({ isDatePickerShown: true });
office-ui-fabric-react:       251 |     const todayButton = document.querySelector('.ms-DatePicker-day--today') as HTMLButtonElement;
office-ui-fabric-react:     > 252 |     ReactTestUtils.Simulate.click(todayButton);
office-ui-fabric-react:           |                             ^
office-ui-fabric-react:       253 | 
office-ui-fabric-react:       254 |     const selectedDate = datePicker
office-ui-fabric-react:       255 |       .find('input')
office-ui-fabric-react: 
office-ui-fabric-react:       at getInstanceFromNode$1 (../../node_modules/react-dom/cjs/react-dom.development.js:888:18)
office-ui-fabric-react:       at Object.click (../../node_modules/react-dom/cjs/react-dom-test-utils.development.js:1209:22)
office-ui-fabric-react:       at Object.<anonymous> (src/components/DatePicker/DatePicker.test.tsx:252:29)
```

**Comment by JasonGore** [comment]
Test should be fixed with the merge of #11856.

File: packages/react-focus/src/components/FocusZone/FocusZone.tsx [hunk 11]
```diff
@@ -427,8 +430,8 @@ export class FocusZone extends React.Component<IFocusZoneProps> implements IFocu
    * Handle global tab presses so that we can patch tabindexes on the fly.
    */
   private _onKeyDownCapture = (ev: KeyboardEvent): void => {
-    if (ev.which === KeyCodes.tab) {
-      _outerZones.forEach(zone => zone._updateTabIndexes());
+    if (ev.key === 'Tab') {
```
Comment (aneeshack4) [inline packages/react-focus/src/components/FocusZone/FocusZone.tsx]: why this change to not use KeyCodes?

Comment (khmakoto) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: `ev.which` and `ev.keyCode` are deprecated:

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/which
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

Comment (aneeshack4) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: Ok would be good to remove the rest of this usage too but can be done in subsequent PRs as well

Comment (ecraig12345) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: What's the recommended alternative?

Comment (ecraig12345) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: Can we update the key codes enum or add a new one so we're not stuck using strings which aren't type safe?

Comment (dzearing) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: wait all. don't make a single code change in this pr. separate ones.

Comment (khmakoto) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: Ok, I'll revert `which` and just add a tslint exception so that the linter doesn't complain. For reference, [these](https://github.com/OfficeDev/office-ui-fabric-react/pull/11765/commits/a34ca1ebef4c5f1248c5fa10e219fa8ce34b45d2) are the only things I changed.

Comment (dzearing) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: Here's what I was thinking:

Step 1. Single PR. Break out FZ into package. Don't remove old.
Step 2. Get it deployed and published. Make sure it's consumable and safe.
Step 3. Separate PR, now make OUFR depend on it.

I think we might be doing too many steps here.

Comment (khmakoto) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: For sure, I'll remove what I did and just add a rule at the top to disable the linter in that file for now, since it won't pass the build if I don't disable it.

Comment (dzearing) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: BTW @khmakoto ev.key doesn't seem to be supported in android browser.

https://caniuse.com/#search=KeyboardEvent.key

Comment (dzearing) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: Also, we should adopt this eventually https://github.com/levithomason/keyboard-key/blob/master/src/keyboardKey.js

Comment (dzearing) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: Let's get the package to a v1 state, so that we can ship it in current without breaking everyone, then remove deprecations and stage that for a next release.

File: packages/office-ui-fabric-react/package.json [hunk 12]
```diff
@@ -71,6 +71,7 @@
     "sinon": "^4.1.3"
   },
   "dependencies": {
+    "@fluentui/react-focus": "^0.1.0",
```
Comment (dzearing) [inline packages/office-ui-fabric-react/package.json]: Can we not do this dependency in production? The docs literally say: no production code should use a v0.

Let's stage this. Leave the old code. Publish. Then in a separate PR do this and validate that nothing breaks.

**Comment by khmakoto** [comment]
@dzearing, @ecraig12345 I've now left a copy of `FocusZone` inside the oufr package until we publish the `react-focus` one. I think this PR should be ready to be merged without contention as nothing official is dependent on `react-focus` now.

File: packages/office-ui-fabric-react/package.json (hunk 12 above)
Comment (khmakoto) [reply packages/office-ui-fabric-react/package.json]: Done!

File: packages/react-focus/src/components/FocusZone/FocusZone.tsx (hunk 11 above)
Comment (khmakoto) [reply packages/react-focus/src/components/FocusZone/FocusZone.tsx]: Reverted the deprecation-related changes!

File: packages/react-focus/package.json [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    "@types/jest": "~24.9.0",
+    "@types/react": "16.8.11",
+    "@types/react-dom": "16.8.4",
+    "@types/react-test-renderer": "^16.0.0",
+    "@types/webpack-env": "1.13.9",
+    "@uifabric/build": "^7.0.0",
+    "@uifabric/jest-serializer-merge-styles": "^7.0.6",
+    "@uifabric/prettier-rules": "^7.0.4",
+    "@uifabric/tslint-rules": "^7.0.5",
+    "enzyme": "~3.9.0",
```
Comment (JasonGore) [inline packages/react-focus/package.json]: I think this got out of sync with other enzyme versions in the repo, causing master to fail due to a required yarn.lock update.
